### ISSUE TITLE
Observable ES6 Map & Set

### ISSUE THREAD

**Issue by Strate** [mobxjs/mobx#632]
As you know, ES6 provide a set of useful collections, such as `Map` and `Set`, It would be great if mobx will have full support to make them observables.

**Comment by tcrognon** [comment]
I want this very badly, too. The reason this isn't implemented yet is because it requires ES6 Proxy to work. I'm guessing everything will be rewritten to use proxies eventually. For now, map() and asMap() exist because of this limitation.

**Comment by Strate** [comment]
> The reason this isn't implemented yet is because it requires ES6 Proxy to work. 
For what case? Why don't create own object, which looks like `Map` or `Set`, but observable? Such as `ObservableArray` done.

**Comment by urugator** [comment]
@Strate There is a [Map](https://mobxjs.github.io/mobx/refguide/map.html) (only string keys are supported), no `Set` however.
Proxies or extendable built-ins are needed just for built-in type preservation.

**Comment by Strate** [comment]
@urugator just for claryfication: we need es6 proxy support or extendable builtins only for `instanceof` operator, yes?

**Comment by urugator** [comment]
@Strade afaik yes (`Array.isArray` in case of arrays). 
Theoretically built-ins could also provide better performance.
Proxies are generally attractive, because they could allow to keep the original object intact and move the observable logic outside. However I don't see proxies reliably usable in browsers anytime soon.

**Comment by Strate** [comment]
@urugator proxy available at almost all browsers (http://caniuse.com/#search=proxy)
Also we can break `instanceof` operator for that, or just inherit original `Map` & `Set` constructors to get `instanceof` works.
More genrally, I think it could be useful to hook into mobx to convert any value to observable with custom logic. For example, I use my own implementation of collections (replacement for arrays, map & sets), and for me that functionality would be useful.

**Comment by andykog** [comment]
What is the necessity of the `instanceof Map` check? I think, it would be just fine to allow using anything as `ObservableMap` key, in other aspects its seems to comply with standard `Map`. Don't like the idea of using proxies for now, there are many people who need ie9+ support.

**Comment by urugator** [comment]
I agree with @andykog, but I would still appreciate dedicated `ObservableSet`

**Comment by mweststrate** [comment]
The current observableMaps primarily usecase is providing an alternative to dynamically keyed observable object. For that reason it is string based and doesn't require a Map (polyfill), but works on all ES5 environments.

For that reason it might be interesting to be able have separate concept that makes ES6 Maps / Sets observable. I think this could done for example by having function in `mobx-utils` that monkey patch ES6 Maps / Sets etc to be observable. The implementation would probably be similar to the current maps, but using the original map as backing data :)

If anybody likes to implement this, please let me know! Implementation could be along the current lines of ObservableMap. But by either monkey patching or creating a new collection that inherits from `Map`, `instanceof` checks etc would work as expected.

**Comment by Strate** [comment]
@mweststrate I suggest to add it ti the core, to achieve automatic deep observable, as iy done with arrays:

```
cosnt store = observable({
  items: [], // array here become ObservableArray
  itemsHash: new Map() // this map should be ObservableMap
  itemsSet: new Set() // this set should be ObservableSet
})
```

**Comment by mweststrate** [comment]
MobX 3 will automatically convert (string keyed) ES 6 maps to observable maps, just like arrays. For observable sets, see #69. So closing this issue for now.

**Comment by damonmaria** [comment]
@mweststrate `@observable x = new Map()` was working for me with 3.1.0. But now (on 3.1.7) it doesn't (`x` stays an ES6 `Map`, is not converted to an `observable.map`). Is this the intention? Maps are mentioned in the recent changelog but doesn't directly refer to this.

**Comment by nicodjimenez** [comment]
^^^ @damonmaria thank you! And thank you so much for your heroic effort on this project @mweststrate!

**Comment by solkimicreb** [comment]
Most Built-ins (Map and Set for example) won't work with Proxies. Their internal slots expect the Map / Set instance as the `this`, but they receive the Proxy object and throw an `invalid receiver error` on method invocations. The only viable solution I found for this is monkey patching the Map / Set methods (luckily this can be done) with reactivity logic. Arrays can be wrapped with Proxies, but have a somewhat strange behavior with length changes.

**Comment by mweststrate** [comment]
Thanks!

Op za 19 aug. 2017 07:08 schreef Miklos Bertalan <notifications@github.com>:

**Comment by urugator** [comment]
@solkimicreb Can you please help me to replicate the problem? I did a very trivial test in node (not in browser) and it seems ok:
```javascript
let m = new Map();
let p = new Proxy(m, {
  get(map, prop, proxy) {        
    let result = Reflect.get(map, prop, map);            
    if (typeof result === "function") {      
      result = result.bind(map);
    }    
    return result;
  }
});
p.set("a", "aVal");   
console.log(p.has("a"));
console.log(p.get("a"));
console.log(p.size);
p.forEach((val, key) => {
  console.log(key, val);
});
```

**Comment by solkimicreb** [comment]
you are binding to returned methods to the map before returning them so it
wont throw an error. the proxy wont do much in this case though. the
returned function uses the raw map as the this context so it wont intrcept
operations done internally by the map methods. alas you still have to patch
them. a combination of proxy and monkey patching can be used to intercept
builtin methods and non standard expando props (for metadata for example)
on maps and sets. i dont have my laptop with me now, i can write a proper
answer with code in a few hours. sorry for the potato spelling and
formatting

**Comment by urugator** [comment]
@solkimicreb I think I understand - lets say the map call somewhere internally `this.get(key)` the call won't be intercepted, therefore no subscription can occur. But I am not sure where is it neccessary.
I think that Map api has basically two types of subscriptions - subscribe for any change (entries/forEach/...) and subscribe for a specific key change (has/get). Maybe a third kind for keys change only.
In other words ... in which map method, we can't say which elements of the map will (or will not) be accessed internally in advance?

**Comment by solkimicreb** [comment]
@urugator Your are correct, two keys should be enough. I mean that Proxies add no value in this case.

You still have to monkey patch the methods to determine when they are called (and when to register subscriptions). The Proxy simply provides information about getting the methods and not about invoking them. It adds a slight overhead by the traps and a pretty big one by a new bound function on every method access.

Proxy would be nice if it could correctly intercept internals for builtins - as it would allow them to be handled like normal object - but this is not the case. The use case for Proxies for Maps and Sets are to allow users to use them like POJOs - by bypassing their standard API - which is not a pattern to be encouraged anyways.

**Comment by urugator** [comment]
@solkimicreb

>  a new bound function on every method access.

I think you can create the functions only once at proxy creation:
```javascript
function get(key) {
  console.log(`SUBSCRIBE FOR ${key}`);  
  return this.get(key);
}
let m = new Map();
const boundGet = get.bind(m);
let p = new Proxy(m, {
  get(map, prop, proxy) {        
    if (prop === "get") {
      return boundGet;
    }    
  }
});
m.set("a", "aVal");
console.log(p.get("a")); 
```

**Comment by solkimicreb** [comment]
@urugator I am not understood your point. So here is my issue in whole.

The base idea of transparent reactivity is this:

1. Detect if a function uses a property.
2. Call that function automatically if the property changes later.

For this we have to know when a property is accessed and mutated.

Map props are accessed and mutated by standard methods. These methods can be thought of as boxes that need to be cracked to intercept the access / mutate operation inside them. We have to know when a property is accessed or mutated and we have to know which one it was to reqister or trigger the appropriate reactions.

In your example you replace the box with another sealed box, that still can't provide information about when and with what arguments it was called. Monkey patching is inevitable to crack the box in my opinion.

Where and how would you register subscriptions in your examples?

**Comment by urugator** [comment]
@solkimicreb

I tried to create a quick naive prototype and came up with something partially working. I realized that you have to subscribe for keys which have been accessed, but don't exist yet ([see this](https://github.com/mobxjs/mobx/blob/master/src/types/observablemap.ts#L91)), which kinda complicates the whole thing, so that's not supported. I haven't bothered with methods returning iterator.
Also traps should be probably wrapped in actions, but it breaks `forEach` not sure why... I bet there are other problems as well, but here is the code just to give the idea:  

```javascript
const Mobx = require("mobx");

let m = new Map();
// Creates traps for specific map
function createTraps(map) {
  const anyChangedAtom = new Mobx.Atom("AnyChanged");
  const allChangedAtom = new Mobx.Atom("AllChanged");
  const mapHas = map.has.bind(map);
  const mapGet = map.get.bind(map);
  const mapSet = map.set.bind(map);
  const mapClear = map.clear.bind(map);
  const mapDelete = map.delete.bind(map);
  const mapForEach = map.forEach.bind(map);
  const traps = {
    get(key) {      
      allChangedAtom.reportObserved(); // subscribes for operations which affects all elements
      if (mapHas(key)) {
        return mapGet(key).get(); // subscribes for get(key)
      }
      return undefined; // WRONG we have to subscribe in advance for this key
    },
    set(key, value) {
      if (mapHas(key)) {
        mapGet(key).set(value); // reports change for get(key)        
      } else {
        mapSet(key, Mobx.observable.box(value));
      }
      anyChangedAtom.reportChanged();
    },
    has(key) {
      if (mapHas(key)) {
        return mapGet(key).get(); // reports change for get(key) 
      } else {
        return undefined; // WRONG we have to subscribe in advance for this key
      }
      allChangedAtom.reportObserved();      
    },
    delete(key) {      
      if (mapHas(key)) {
        const box = mapGet(key); 
        mapDelete(key); 
        box.set(undefined); // reports change for get(key)
      }      
      anyChangedAtom.reportChanged(); // report to forEach
    },
    clear() {
      mapClear();
      anyChangedAtom.reportChanged();
      allChangedAtom.reportChanged();      
    },
    forEach(callback, thisArg) { // all keys and values are accessed so we want to subscribe for any change      
      mapForEach(callback, thisArg);
      anyChangedAtom.reportObserved();
    }
  }  
  return traps;
}
//
const traps = createTraps(m);
let p = new Proxy(m, {
  get(map, prop, proxy) {
     return traps[prop];
  }
});
// Subscribed for specific key
console.log(["TEST get()","EXPECTING:", "aVal", "aVal1", "aVal2", "undefined", "GOT:"].join("\n"));
p.set("a", "aVal"); // must be created in advance :(
let dispose = Mobx.autorun(() => {  
  console.log(p.get("a"));  
})
p.set("a", "aVal1");
p.set("b", "bVal"); // does not affect autorun
p.set("a", "aVal2");
p.delete("b"); // does not affect autorun
p.delete("a");
//p.clear();
dispose()
// ForEach affected by everything
console.log(["TEST forEach()","EXPECTING:", "1", "2", "3", "4", "5", "GOT:"].join("\n"));
let it = 0; 
dispose = Mobx.autorun(() => {  
  console.log(++it);
  p.forEach((value, key) => {})
  
})
p.set("a", "aVal1");
p.set("b", "bVal1");
p.delete("a");
p.clear();
dispose();
```

**Comment by solkimicreb** [comment]
That looks perfectly functional, I think we are speaking about the same thing here. I would still consider changing this:

```
const traps = createTraps(m);
let p = new Proxy(m, {
  get(map, prop, proxy) {
     return traps[prop];
  }
});
```

to this though.

```
const traps = createTraps(m);
for (let method in m) {
  m[method] = traps[method];
}
// Side note: be sure to overwrite the `size` accessors and `Symbol.iterator` too.
```

Proxies add the benefit of intercepting new props or operations other than get / set. Neither is needed here.

Proxies could provide 2 benefits here though.

- It could allow subscriptions for non standard map extensions (like custom props on them).
- It could allow a Proxy like transparent wrapping. With simple monkey patching you always expose the patched (reactive object) and you can't access a non reactive counterpart. With Proxies you can decide if you would like to expose the patched reactive or the original method on access. (Which aligns better with the transparent wrapping philosophy).

**Comment by nestharus** [comment]
Hey guys,

The company I work for ended up needing ObservableMap to behave like an es6 map. Here is ObservableMap with most mobx functionality (broke whatever wasn't proper es6 behavior).

I didn't want to edit the source, so I just rewrote a bunch of the mobx utility functions.

	import {
		Atom,
		observable,
		extras,
		isObservable,
		isObservableMap,
		isObservableArray,
		transaction,
		untracked
	} from 'mobx';

	import * as mobx from 'mobx';

	function hasListeners(object) {
		return object.changeListeners && object.changeListeners.length > 0;
	}

	function notifyListeners(object, change) {
		if (!object.changeListeners) { return; }
		
		for (let listener of object.changeListeners) {
			listener(change);
		}
	}

	function hasInterceptors(object) {
		return object.interceptors && object.interceptors.length > 0;
	}

	function interceptChange(object, change) {
		if (!object.interceptors) { return; }
		
		for (let interceptor of object.interceptors) {
			change = interceptor(change);
			
			if (change && typeof change !== 'object') {
				throw new Error('Intercept handlers should return nothing or a change object');
			}
			
			if (!change) {
				break;
			}
		}
		
		return change;
	}

	function isPlainObject(value) {
		if (value === undefined || typeof value !== 'object') return false;
		const proto = Object.getPrototypeOf(value);
		return proto === Object.prototype || proto === null;
	}

	function isES6Map(value) {
		return value instanceof Map;
	}

	function referenceEnhancer(value) {
		return value;
	}

	function checkIfStateModificationsAreAllowed(atom) {
		const hasObservers = atom.observers.length > 0;
		
		if (extras.getGlobalState().computationDepth > 0 && hasObservers) {
			throw new Error("Computed values are not allowed to cause side effects by changing observables that are already being observed. Tried to modify: " + atom.name);
		}
		
		if (!extras.getGlobalState().allowStateChanges && hasObservers) {
			throw new Error((extras.getGlobalState().strictMode?
				"Since strict-mode is enabled, changing observed observable values outside actions is not allowed. Please wrap the code in an `action` if this change is intended. Tried to modify: "
				:
				"Side effects like changing state are not allowed at this point. Are you trying to modify state from, for example, the render function of a React component? Tried to modify: ")
				+ atom.name);
		}
	}


	export class ObservableMap extends mobx.ObservableMap {
		constructor(keys, enhancer, name) {
			super(undefined, enhancer, name);
			
			let iterable;
			if (Array.isArray(keys) || isObservableArray(keys)) {
				iterable = keys;
			}
			else if (isES6Map(keys) || isObservableMap(keys)) {
				iterable = [];
				
				for (let [key, value] of keys.entries()) {
					iterable.push([key, value]);
				}
			}
			else if (isPlainObject(keys)) {
				iterable = [];
				
				for (let key of Object.keys(keys)) {
					iterable.push([
						key, keys[key]
					]);
				}
			}
			else if (keys !== null && keys !== undefined) {
				throw new Error('Cannot initialize map from ' + keys);
			}
			
			if (iterable !== undefined) {
				for (let i = 0; i < iterable.length; ++i) {
					iterable[i] = this.enhancer(iterable[i]);
				}
			}
			
			const data = new Map(iterable);
			const atoms = new Map();
			const sizeAtom = new Atom();
			const atomAny = new Atom();
			const atomAll = new Atom();
			
			const getAtom = function(key) {
				let atom = atoms.get(key);
				
				if (atom === undefined) {
					atom = {
						has: new Atom(),
						write: new Atom()
					};
					
					atoms.set(key, atom);
				}
				
				return atom;
			};
			
			if (iterable !== undefined && typeof iterable === 'object' && typeof iterable[Symbol.iterator] === 'function') {
				for (let element of iterable) {
					atoms.set(element[0], {
						has: new Atom(),
						write: new Atom()
					});
				}
			}
			
			const executeListeners = (key, oldValue, newValue, changeType) => {
				checkIfStateModificationsAreAllowed(atomAll);
				
				const notifySpy = extras.isSpyEnabled();
				const notify = hasListeners(this);
				
				const change = notify || notifySpy? {
					type: changeType,
					object: this,
					name: key,
					newValue: newValue,
					oldValue: oldValue
				} : undefined;
				
				if (notifySpy) {
					extras.spyReportStart(change);
				}
				
				if (changeType === 'delete') {
					data.delete(key);
				}
				else {
					data.set(key, newValue);
				}
				
				const atom = getAtom(key);
				
				transaction(() => {
					if (changeType !== 'update') {
						atom.has.reportChanged();
						sizeAtom.reportChanged();
					}
					
					atom.write.reportChanged();
					atomAny.reportChanged();
				});
				
				if (notify) {
					notifyListeners(this, change);
				}
				
				if (notifySpy) {
					extras.spyReportEnd();
				}
			};
			
			const updateKey = function(key, value) {
				const oldValue = data.get(key);
				
				if (value !== oldValue) {
					executeListeners(key, oldValue, value, 'update');
				}
			};
			
			const addKey = function(key, value) {
				executeListeners(key, undefined, value, 'add');
			};
			
			const deleteKey = function(key) {
				if (!data.has(key)) {
					return false;
				}
				
				executeListeners(key, data.get(key), undefined, 'delete');
				
				return true;
			};
			
			this._getAtom = getAtom;
			this._anyAtom = atomAny;
			this._sizeAtom = sizeAtom;
			
			this.has = (key) => {
				getAtom(key).has.reportObserved();
				atomAll.reportObserved();
				
				return data.has(key);
			};
			
			this.get = (key) => {
				const atom = getAtom(key);
				
				atom.has.reportObserved();
				atomAll.reportObserved();
				
				if (!data.has(key)) {
					return undefined;
				}
				
				atom.write.reportObserved();
				
				return data.get(key);
			};
			
			this.set = (key, value) => {
				const hasKey = data.has(key);
				
				if (hasInterceptors(this)) {
					const change = interceptChange(this, {
						type: hasKey? 'update' : 'add',
						object: this,
						newValue: value,
						name: key
					});
					
					if (!change) {
						return this;
					}
					
					value = change.newValue;
				}
				
				if (data.get(key) === value) {
					return this;
				}
				
				value = this.enhancer(value, undefined, undefined);
				
				if (hasKey) {
					updateKey(key, value);
				}
				else {
					addKey(key, value);
				}
				
				return this;
			};
			
			this.delete = (key) => {
				if (hasInterceptors(this)) {
					const change = interceptChange(this, {
						type: 'delete',
						object: this,
						name: key
					});
					
					if (!change) {
						return false;
					}
				}
				
				return deleteKey(key);
			};
			
			this.clear = () => {
				transaction(() => {
					untracked(() => {
						[...data.keys()].forEach(this.delete, this);
					});
				});
			};
			
			this.toString = () => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return this.name + '[{ ' + [...data.keys()].map(key => `${key}: ${'' + data.get(key)}`).join(', ') + ' }]';
			};
			
			this.toJS = () => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return data;
			};
			
			this.toJSON = () => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return this.toJS();
			};
			
			this.replace = (map) => {
				const missingKeys = [];
				
				if (isES6Map(map) || isObservableMap(map)) {
					for (let key of data.keys()) {
						if (!map.has(key)) {
							missingKeys.push(key);
						}
					}
				}
				else if (Array.isArray(map)|| isPlainObject(map) || isObservableArray(map)) {
					for (let key of data.keys()) {
						if (map[key] === undefined) {
							missingKeys.push(key);
						}
					}
				}
				else {
					throw new Error('Must replace map with another map, an object, or an array');
				}
				
				transaction(() => {
					for (let key of missingKeys) {
						this.delete(key);
					}
					
					this.merge(map);
				});
				
				return this;
			};
			
			this.merge = (other) => {
				if (isObservable(other)) {
					other = other.toJS();
				}
				
				transaction(() => {
					if (Array.isArray(other)) {
						other.forEach(([key, value]) => this.set(key, value));
					}
					else if (isES6Map(other)) {
						other.forEach((value, key) => this.set(key, value));
					}
					else if (isPlainObject(other)) {
						Object.keys(other).forEach(key => this.set(key, other[key]));
					}
					else if (other !== null && other !== undefined) {
						throw new Error('Cannot initialize map from ' + other);
					}
				});
				
				return this;
			};
			
			this.forEach = (callback) => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return data.forEach(callback);
			};
			
			this.entries = () => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return data.entries();
			};
			
			this.values = () => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return data.values();
			};
			
			this.keys = () => {
				atomAny.reportObserved();
				atomAll.reportObserved();
				
				return data.keys();
			};
			
			Object.defineProperty(this, 'size', observable({
				get() {
					sizeAtom.reportObserved();
					atomAll.reportObserved();
					
					return data.size;
				}
			}));
			
			return this;
		};
	}

	export class ShallowObservableMap extends ObservableMap {
		constructor(keys, name) {
			super(keys, referenceEnhancer, name);
		}
	}
